### PR TITLE
Feature: OSF Email Notifications

### PR DESCRIPTION
**Pull request by samanehsan**
<h3>Purpose:</h3>
Send emails to users when a comment happens on one of their projects or when another user replies to one of their comments on any OSF project. For each notification event on each project, users will have the option to receive emails instantly (transactional emails) or to include the notification in a daily email digest.

<h3>TODO:</h3> 
- [x] Integrate html email templates by @caneruguz 
- [x] Finish writing tests
- [x] Fix line height/placement of event titles in treebeard
- [x] Refactor logic used in multiple locations into separate functions:
 - [x] getting all user subscriptions for a single node 
 - [x] removing user from a subscription
- [x] Use celery for sending emails

### CODE REVIEW

**Comment by samanehsan** [comment]
This is ready for review @sloria :snowman:

**Comment by lbanner** [comment]
Excellent! Congrats! :smile_cat:

File: website/notifications/views.py [hunk 1]
```diff
@@ -0,0 +1,51 @@
+from framework.auth.decorators import must_be_logged_in
+from model import Subscription
+from flask import request
+from modularodm import Q
+from modularodm.exceptions import NoResultsFound
+from modularodm.storage.mongostorage import KeyExistsException
+from website import settings
+
+@must_be_logged_in
+def subscribe(auth):
```
Comment (sloria) [inline website/notifications/views.py]: Since the name of the a view function determines the key for reverse lookup with `url_for` we may want to name this something more specific. Something `set_subscription` or `configure_subscription`? Look around the existing routes to find a name that is consistent.

File: website/notifications/views.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+    user = auth.user
+    subscription = request.json
+    event = subscription.get('event')
```
Comment (sloria) [inline website/notifications/views.py]: What if `event` and `notification_type` are not provided by the client? Should we return an error response?

File: website/routes.py [hunk 3]
```diff
@@ -1227,6 +1228,13 @@ def make_url_map(app):
         ),
 
         Rule(
+            '/settings/subscribe/',
```
Comment (sloria) [inline website/routes.py]: Subscriptions can be considered a resource, so a more appropriate URL may be `/api/v1/subscriptions/`. 

```
GET /api/v1/subscriptions/ -> return list of subscriptions
POST /api/v1/subscriptions/ -> create a new subscription
PUT /api/v1/subscriptions/<event>/ -> modify an existing subscription to `<event>`
```

File: website/static/css/site.css [hunk 4]
```diff
@@ -2109,3 +2109,46 @@ span#profileFullname{
 .modal-backdrop {
     position: fixed;
 }
+
+/* Treebeard in Configure Notifications, project settings page */
+#notificationSettings #tb-tbody {
```
Comment (sloria) [inline website/static/css/site.css]: There is a lot of duplicated code here. Also, these rules are rather brittle--if we change any of these CSS IDs, the UI breaks.

Could you apply a helper class instead, like `osf-treebeard-minimal`, then refactor this CSS?

```CSS
.osf-treebeard-minimal {
    border: none;
}

.osf-treebeard-minimal .tb-head {
    display: none;
}

...
```

File: website/static/js/osfNotifications.js [hunk 5]
```diff
@@ -0,0 +1,7 @@
+var $ = require('jquery');
```
Comment (sloria) [inline website/static/js/osfNotifications.js]: Remove unused imports.

File: website/static/js/osfNotifications.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
...
+var $osf = require('osfHelpers');
+var bootbox = require('bootbox');
+
+//initialize treebeard
+var ProjectNotifications = require('./project-settings-treebeard.js');
```
Comment (sloria) [inline website/static/js/osfNotifications.js]: This is a page module. It should go in `website/static/js/pages/osf-notifications.js`. 

See http://cosdev.readthedocs.org/en/latest/osf/js_modules.html

File: website/static/js/project-settings-treebeard.js [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+var $osf = require('osfHelpers');
+var bootbox = require('bootbox');
+require('../vendor/bower_components/slickgrid/lib/jquery.event.drag-2.2.js');
+var m = require('mithril');
+var Treebeard = require('treebeard');
+require('../css/fangorn.css');
+
+
+function resolveToggle(item) {
+        var toggleMinus = m('i.icon-minus', ' '),
```
Comment (sloria) [inline website/static/js/project-settings-treebeard.js]: Fix indentation here.

File: website/static/js/project-settings-treebeard.js [hunk 8]
```diff
@@ -0,0 +1,244 @@
...
+            return togglePlus;
+        }
+        item.open = true;
+        return '';
+    }
+
+function resolveIcon(item) {
+    if (item.children.length > 0) {
+        if (item.open) {
+            return m("i.icon.icon-folder-open", " ");
```
Comment (sloria) [inline website/static/js/project-settings-treebeard.js]: Use singlequotes for strings.

File: website/static/js/project-settings-treebeard.js [hunk 9]
```diff
@@ -0,0 +1,244 @@
...
+    var payload = {
+                'id': id,
+                'event': event,
+                'notification_type': notification_type
+    };
+    $osf.postJSON(
+        '/api/v1/settings/subscribe/',
+        payload
+    ).fail(function() {
+        bootbox.alert('Could not update notification preferences.')
```
Comment (sloria) [inline website/static/js/project-settings-treebeard.js]: Missing semicolon.

File: website/static/js/project-settings-treebeard.js [hunk 10]
```diff
@@ -0,0 +1,244 @@
+var $ = require('jquery');
```
Comment (sloria) [inline website/static/js/project-settings-treebeard.js]: Missing `'use strict';`

File: website/static/js/project-settings-treebeard.js [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+    var parent = item.parent();
+    if(parent && parent.id > 0) {
+        tb.updateFolder(null, parent);
+        openAncestors(tb, parent);
+    }
+}
+
+function subscribe(id, event, notification_type) {
+    var payload = {
+                'id': id,
```
Comment (sloria) [inline website/static/js/project-settings-treebeard.js]: Fix indentation.

File: website/static/js/project-settings-treebeard.js [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+                'event': event,
+                'notification_type': notification_type
+    };
+    $osf.postJSON(
+        '/api/v1/settings/subscribe/',
```
Comment (sloria) [inline website/static/js/project-settings-treebeard.js]: See previous comment re: routes.

File: website/static/js/project-settings-treebeard.js [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+                },
+                {
+                    title: "Notification Type",
+                    width : "40%",
+                    sort : false
+
+                }
+            ]},
+        resolveRows : function notificationResolveRows(item){
+            var default_columns = [];
```
Comment (sloria) [inline website/static/js/project-settings-treebeard.js]: Use `camelCase` for variable names. Also, `default_columns` isn't really an appropriate name for this variable (it's not a "default"). It is the return value of this function, which is an array of objects describing columns. So maybe `columns`, or `ret`?

File: website/templates/profile/notifications.mako [hunk 14]
```diff
@@ -44,20 +45,31 @@
...
 </%def>
 
 <%def name="javascript()">
     <% import website %>
+    <% import json %>
     ${parent.javascript()}
     <script type="text/javascript">
-        window.contextVars = $.extend({}, window.contextVars, {'mailingList': '${website.settings.MAILCHIMP_GENERAL_LIST}'});
+        window.contextVars = $.extend({}, window.contextVars, {'mailingList': '${website.settings.MAILCHIMP_GENERAL_LIST}',
+                                                               'notificationsData': ${json.dumps(notifications_data)}
```
Comment (sloria) [inline website/templates/profile/notifications.mako]: Once you've added an endpoint to `GET /api/v/subscriptions/` (see previous comment), this should no longer be necessary.

File: website/templates/profile/notifications.mako [hunk 15]
```diff
@@ -44,20 +45,31 @@
...
+        });
+
     </script>
 </%def>
 
 <%def name="javascript_bottom()">
     ${parent.javascript_bottom()}
     <script src="${"/static/public/js/notifications-config-page.js" | webpack_asset}"></script>
+    <script src="${"/static/public/js/osf-notifications-config-page.js" | webpack_asset}"></script>
```
Comment (sloria) [inline website/templates/profile/notifications.mako]: Why do we need to page modules for notifications?

File: website/util/send_digest.py [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+                                    'timestamp': curr.timestamp
+                                },
+                                'node_lineage': curr.node_lineage
+                            }
+                            result.info.push(info);
+                    };
+                    """))
+
+
+class TestSendDigest(OsfTestCase):
```
Comment (sloria) [inline website/util/send_digest.py]: Move this to `tests/test_utils.py`.

File: website/project/views/node.py [hunk 17]
```diff
@@ -328,8 +328,11 @@ def node_setting(auth, **kwargs):
         'level': node.comment_level,
     }
 
+    ret['subscriptions'] = format_data(auth.user, [node._id], [])
```
Comment (sloria) [inline website/project/views/node.py]: This, too, will be unnecessary if you expose an endpoint to GET subscriptions.

File: website/notifications/utils.py [hunk 18]
```diff
@@ -0,0 +1,160 @@
...
+from website import settings
+from website.models import Node
+from website.notifications.model import Subscription
+from modularodm import Q
+from modularodm.exceptions import NoResultsFound
+
+
+class NotificationsDict(dict):
+    def __init__(self):
+        super(dict, self).__init__()
```
Comment (sloria) [inline website/notifications/utils.py]: This should be either

File: website/notifications/utils.py [hunk 19]
```diff
@@ -0,0 +1,160 @@
...
+        user_subscriptions = get_all_user_subscriptions(user)
+    node_subscriptions = []
+    for s in user_subscriptions:
+        if s.object_id == node._id:
+            node_subscriptions.append(s)
+
+    return node_subscriptions
+
+
+def format_data(user, node_ids, data, subscriptions_available=settings.SUBSCRIPTIONS_AVAILABLE):
```
Comment (sloria) [inline website/notifications/utils.py]: Can you break this into smaller functions? Particularly, the logic to serialize nodes and events can be separated. This can go a long way to make the code easier to read, maintain, and test.

File: website/notifications/utils.py [hunk 20]
```diff
@@ -0,0 +1,160 @@
...
+    user_subscriptions = get_all_user_subscriptions(user)
+    for idx, node_id in enumerate(node_ids):
+        node = Node.load(node_id)
+        index = len(data)
+        data.append({'node_id': node_id,
```
Comment (sloria) [inline website/notifications/utils.py]: Following [this note from [cosdev](http://cosdev.readthedocs.org/en/latest/osf/guidelines.html#responses): 

> Prefer namespaced representations to arbitrary prefixes 

it'd be preferable to have `{'node': {'id': node_id, 'url': node.url ...}}`

File: website/notifications/utils.py [hunk 21]
```diff
@@ -0,0 +1,160 @@
...
+            for notification_type in settings.NOTIFICATION_TYPES:
+                if user in getattr(subscription, notification_type):
+                    return notification_type
+
+
+def format_user_and_project_subscriptions(user):
+    return [
+        {
+            'title': 'User Notifications',
+            'node_id': user._id,
```
Comment (sloria) [inline website/notifications/utils.py]: See previous comment re: data formatting.

File: website/notifications/views.py [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+
+        if user not in getattr(s, notification_type):
+            getattr(s, notification_type).append(user)
+            s.save()
+
+        for nt in settings.NOTIFICATION_TYPES:
+            if nt != notification_type:
+                if getattr(s, nt) and user in getattr(s, nt):
+                    getattr(s, nt).remove(user)
+                    s.save()
```
Comment (sloria) [inline website/notifications/views.py]: All endpoints should return useful information back to the client, with proper status codes.

File: website/util/send_digest.py (hunk 16 above)
Comment (samanehsan) [inline website/util/send_digest.py]: I thought the convention was to put tests for scripts in the same file as the script?

File: website/settings/defaults.py [hunk 23]
```diff
@@ -74,6 +74,21 @@ def parent_dir(path):
 MAIL_USERNAME = 'osf-smtp'
 MAIL_PASSWORD = ''  # Set this in local.py
 
+# Notification options
+SUBSCRIPTIONS_AVAILABLE = {
```
Comment (sloria) [inline website/settings/defaults.py]: These constants aren't meant to configurable, so they don't really belong in `website.settings`. Maybe move them to `website.notifications.constants`.

File: website/notifications/__init__.py [hunk 24]
```diff
@@ -0,0 +1 @@
+__author__ = 'Saman'
```
Comment (sloria) [inline website/notifications/__init__.py]: We haven't really established a policy for designating authorship/ownership of parts of the codebase. This is a tricky area, and it only gets more complex as a project matures. For now, let's omit any designations of authorship aside from the git history.

File: website/notifications/emails.py [hunk 25]
```diff
@@ -0,0 +1,143 @@
...
+from framework.tasks import app
+from framework.tasks.handlers import queued_task
+from website import mails, settings
+from website.util import web_url_for
+from website.models import Node, User
+from mako.lookup import Template
+
+
+def notify(uid, event, **context):
+    key = str(uid + '_' + event)
```
Comment (sloria) [inline website/notifications/emails.py]: This sort of formatting is an implementation detail that should be hidden. Can you move this to a tiny little helper function `to_subscription_key(uid, event)`.?

File: website/util/send_digest.py (hunk 16 above)
Comment (sloria) [inline website/util/send_digest.py]: That is the convention for one-off scripts such as migrations. Since we will be using this script more-than-once, we will want to make sure that it runs with our CI builds.

File: website/util/send_digest.py [hunk 26]
```diff
@@ -0,0 +1,162 @@
...
+from website.notifications.utils import NotificationsDict
+from tests.base import OsfTestCase
+from tests.factories import DigestNotificationFactory, UserFactory, ProjectFactory
+import nose.tools
+
+
+def main():
+    init_app(routes=False)
+    grouped_digests = group_digest_notifications_by_user()
+    print send_digest(grouped_digests)
```
Comment (sloria) [inline website/util/send_digest.py]: Why `print` (send_digest doesn't return a string)?

**Comment by sloria** [comment]
1st pass finished. :mailbox:

File: website/routes.py (hunk 3 above)
Comment (samanehsan) [inline website/routes.py]: Does this mean I have to separate the logic for creating and updating a subscription into two separate functions? Currently, it's all handled by the `subscribe` function: 
https://github.com/samanehsan/osf.io/blob/feature/notifications/website/notifications/views.py#L10

File: website/notifications/__init__.py (hunk 24 above)
Comment (samanehsan) [inline website/notifications/__init__.py]: Sure! I meant to delete this file anyways.

File: website/util/send_digest.py (hunk 16 above)
Comment (samanehsan) [inline website/util/send_digest.py]: Ok, got it!

File: website/routes.py (hunk 3 above)
Comment (sloria) [inline website/routes.py]: Yes, you will have multiple view functions (one for each verb).

File: website/notifications/__init__.py (hunk 24 above)
Comment (sloria) [inline website/notifications/__init__.py]: You'll still need this file in order to make `website.notifications` a proper module.

File: website/routes.py (hunk 3 above)
Comment (samanehsan) [inline website/routes.py]: For deleting subscriptions, I'm sending a signal when a node is deleted (https://github.com/samanehsan/osf.io/blob/feature/notifications/website/project/model.py#1438). How would I change this to use a `DELETE ` endpoint?

**Comment by samanehsan** [comment]
Done responding to first pass code review @sloria! :email:

**Comment by lbanner** [comment]
Ninja skills! :zap:

**Comment by samanehsan** [comment]
This is ready for another pass @sloria! :snowflake:

File: framework/tasks/__init__.py [hunk 27]
```diff
@@ -10,7 +10,7 @@
 from website import settings
 
 
-app = Celery()
+app = Celery('scripts.send_digest')
```
Comment (JeffSpies) [inline framework/tasks/__init__.py]: Should we be hardcoding this in framework/tasks?

Comment (samanehsan) [inline framework/tasks/__init__.py]: I'll see if I can set this in `scripts.send_digest` instead

File: bower.json [hunk 28]
```diff
@@ -27,13 +27,14 @@
...
+    "treebeard": "https://github.com/caneruguz/treebeard.git#e22caee9ebf91cc00f73e3912883d2d5c09e1300",
     "jquery.cookie": "~1.4.1",
     "hgrid": "~0.2.10",
     "uri.js": "~1.14.1",
     "xhook": "~1.3.0",
     "styles": "https://github.com/lyndsysimon/styles.git#4db9c8e2f5932cad9f23ad3579d1a82c691faf12",
     "locales": "https://github.com/citation-style-language/locales.git#d2b612f8a6f764cbd66e67238636fac3888a7736",
-    "c3": "~0.4.9"
+    "c3": "~0.4.9",
+    "jsTimezoneDetect": "https://cdnjs.cloudflare.com/ajax/libs/jstimezonedetect/1.0.4/jstz.js"
```
Comment (brianjgeiger) [inline bower.json]: Not a huge deal, but is there a way to do this not from a cdn? At least this is pinned to a version, but the less we have to rely on CDNs the better, I think.

Comment (samanehsan) [inline bower.json]: @sloria and I tried to get the library directly from the repo, but it didn't work for some reason

File: scripts/send_digest.py [hunk 29]
```diff
@@ -0,0 +1,112 @@
...
+from website import mails
+from website.util import web_url_for
+from website.app import init_app
+from website.notifications.model import DigestNotification
+from website.notifications.utils import NotificationsDict
+
+from scripts import utils as script_utils
+
+logger = logging.getLogger(__name__)
+logging.basicConfig(level=logging.DEBUG)
```
Comment (erinspace) [inline scripts/send_digest.py]: I remember I was told once to not set this in production, should this logging config be removed?

File: website/templates/profile/notifications.mako [hunk 30]
```diff
@@ -44,13 +45,23 @@
...
-            </div>
+        </div>
+        <div class="panel panel-default">
+            <div class="panel-heading"><h3 class="panel-title">Configure Notification Preferences</h3></div>
+                <form id="selectNotifications" class="osf-treebeard-minimal">
+                    <div id="grid"></div>
+                    <div class="help-block" style="padding-left: 15px">
+                            <p id="configureNotificationsMessage"></p>
+                    </div>
+                </form>
```
Comment (HarryRybacki) [inline website/templates/profile/notifications.mako]: Should we have a submit button here for the user to confirm the setting? Currently, selecting email or email digests from the drop down works but, it's different from all of the other config sections on the page.

Comment (brianjgeiger) [inline website/templates/profile/notifications.mako]: I think some feedback that things have been saved would do the same thing without requiring a submit button (which would likely be difficult to implement at this point). And better in the long run, even if it's not entirely consistent.

Comment (samanehsan) [inline website/templates/profile/notifications.mako]: While it's not consistent with other forms on the page, the advantage of having the settings update automatically is that the user won't loose these changes if they happen to navigate away.

File: scripts/send_digest.py (hunk 29 above)
Comment (samanehsan) [inline scripts/send_digest.py]: Oh, good point! I was using this for debugging, but I don't think we need to see the output of every email sent in the logs.

File: website/templates/profile/notifications.mako (hunk 30 above)
Comment (caneruguz) [inline website/templates/profile/notifications.mako]: Actually it's not very hard to implement. Treebeard has a notify class that is used for this purpose, it shows a message replacing either the entire row or a column and disappears after 2 seconds (or what you set). I can show you how to implement it, but fangorn.js has many examples, like: 
```javascript
        item.notify.update(msgText, 'success', column, 0);
```
It also has good documentation here: 
https://github.com/caneruguz/treebeard/wiki/Notify

Comment (HarryRybacki) [inline website/templates/profile/notifications.mako]: True. However users will lose any setting changes if they were to navigate away before hitting submit on anything else as well. More a matter of consistency than usefullness imo.

File: scripts/send_digest.py [hunk 31]
```diff
@@ -0,0 +1,113 @@
...
+
+
+def main():
+    script_utils.add_file_logger(logger, __file__)
+    app = init_app()
+    celery_app.main = 'scripts.send_digest'
```
Comment (samanehsan) [inline scripts/send_digest.py]: I've named the celery_app in `scripts.send_digest.py` but I think it should be set back to `None` after the script executes so as not to interfere with other celery imports. Would I need to do this in a second callback function? @JeffSpies @sloria

File: website/templates/profile/notifications.mako (hunk 30 above)
Comment (brianjgeiger) [inline website/templates/profile/notifications.mako]: Sorry, I meant that having to use a submit button would be harder than the notification. Ideally, things that can be changed but require submit should have cancel buttons and dirty flags to warn users that they will lose changes. I would implement that (if it's not implemented) and keep this live rather than forcing consistency on a bad user interface pattern, with the ultimate goal of replacing out all of the submit buttons with live updates where possible.

Comment (HarryRybacki) [inline website/templates/profile/notifications.mako]: What if we were to implement it in a consistent manner now and submit a separate PR adjusting the behavior or all sections uniformly after?

Comment (samanehsan) [inline website/templates/profile/notifications.mako]: I agree with BrianG in that since this is probably something that we want to work toward and I've already implemented it, there's no need to revert back to using a submit button only to then change it back to auto-submit at a later date.

Comment (HarryRybacki) [inline website/templates/profile/notifications.mako]: Could we add a growl or something to inform the user that the change has been made? I wasn't sure that it had gone through until I manually tested the feature after the fact.

File: website/notifications/utils.py [hunk 32]
```diff
@@ -0,0 +1,225 @@
...
+        'event': parsed_key[1]
+    }
+
+
+@contributor_removed.connect
+def remove_contributor_from_subscriptions(contributor, node):
+    """ Remove contributor from node subscriptions unless the user is an
+        admin on any of node's parent projects.
+    """
+    if not contributor._id in node.admin_contributor_ids:
```
Comment (brianjgeiger) [inline website/notifications/utils.py]: With the current permission structure, which I haven't looked at the code for but know its effects, shouldn't this resolve to  the user having read permissions? Or is there another reason to exclude admins?

I tried to find the usage of this method in the code so I could understand when it would be used, but it doesn't appear to be used. If it's not being used, it should either be removed, or there should be more explanation in the comment about why admins are not included.

File: website/templates/profile/notifications.mako (hunk 30 above)
Comment (JeffSpies) [inline website/templates/profile/notifications.mako]: With no save button, the assumption should be that it has saved. If we
can't do a little green checkmark next to the input, then let's go
message-less.

Comment (brianjgeiger) [inline website/templates/profile/notifications.mako]: Caner says that in-line notifications are very easy and the way to do it is detailed above.

Comment (samanehsan) [inline website/templates/profile/notifications.mako]: I think an inline message with treebeard might be the way to go then. If I'm not mistaken, the growl notifications stack up so if a user configures several notifications at once they would have to manually close them all.

Comment (brianjgeiger) [inline website/templates/profile/notifications.mako]: Growlboxes can be set to disappear after a time, but inline would be better in any case.

Comment (JeffSpies) [inline website/templates/profile/notifications.mako]: Yeah, growl doesn't make sense here, although I respect Harry's desire for
some message. Let's do inline. Maybe Harry can hack that up.

Comment (caneruguz) [inline website/templates/profile/notifications.mako]: Harry if you need help, like I said before, there is documentation and examples of this: 
https://github.com/caneruguz/treebeard/wiki/Notify

And I'm available if you have any questions.

File: website/notifications/utils.py (hunk 32 above)
Comment (samanehsan) [inline website/notifications/utils.py]: The `Node.remove_contributor` "calls" this function when it sends a `contributor_removed` signal, which is then received by this function. 

Admins with read-permissions can still subscribe to notifications, just like read-only contributor can. 

They are excluded in the case where they are removed from a component as a contributor, but they are an admin on the parent project. This is because the admin still has read-access to the component and thus should still get notifications.

File: website/templates/profile/notifications.mako (hunk 30 above)
Comment (HarryRybacki) [inline website/templates/profile/notifications.mako]: I'll work on it :)

**Comment by HarryRybacki** [comment]
@samanehsan @JeffSpies the conditional [here](https://github.com/samanehsan/osf.io/blob/feature/notifications/website/templates/profile/addons.mako#L16-L18) and [here](https://github.com/samanehsan/osf.io/blob/feature/notifications/website/templates/profile/account.mako#L16-L18) should be reworked. Currently it makes developing locally difficult as this flag needs to be true in order to work with notifications on the user settings page. However, if the flag is true than the local dev environment also needs Mailchimp API keys set.

File: website/notifications/emails.py [hunk 33]
```diff
@@ -0,0 +1,170 @@
...
+
+def get_settings_url(uid, user):
+    if uid == user._id:
+        return web_url_for('user_notifications', _absolute=True)
+    else:
+        return Node.load(uid).absolute_url + 'settings/'
+
+
+def localize_timestamp(timestamp, user):
+    user_timezone = pytz.timezone(user.timezone)
```
Comment (brianjgeiger) [inline website/notifications/emails.py]: The chain of timezone usage relies on a 3rd party javascript extension. While it will probably work fine, there's nothing yet set to handle if it, for example, were to say that the timezone is set to `''`. If that happens, you will get an UnknownTimezoneError here. Probably catching it here and doing something sane would be the ideal thing to do (for example, not trying to localize the time but assuming GMT). Also, if it received a `None` here, you'd get an AttributeError. I don't know if it's possible to get a `None`, but an empty string or something equally invalid in string form seems very possible.

File: tests/test_views.py [hunk 34]
```diff
@@ -3509,6 +3509,14 @@ def test_folder_new_post(self):
...
         assert_true(found_item, "Did not find the folder in the dashboard.")
 
+    def test_update_user_timezone_offset(self):
+        assert_equal(self.creator.timezone, 'Etc/UTC')
+        payload = {'timezone': 'America/New_York'}
+        url = api_url_for('update_user', uid=self.creator._id)
+        self.app.put_json(url, payload, auth=self.creator.auth)
+        self.creator.reload()
+        assert_equal(self.creator.timezone, 'America/New_York')
+
```
Comment (HarryRybacki) [inline tests/test_views.py]: As per @brianjgeiger's comment regarding the JS library we are using for this; should we also test for handling improper timezone submissions?

File: website/notifications/model.py [hunk 35]
```diff
@@ -0,0 +1,28 @@
+from modularodm import fields
+from framework.mongo import StoredObject, ObjectId
+from website.notifications.constants import NOTIFICATION_TYPES
+
+
+class Subscription(StoredObject):
+    _id = fields.StringField(primary=True)  # pxyz_wiki_updated, uabc_comment_replies
```
Comment (HarryRybacki) [inline website/notifications/model.py]: It isn't readily clear to me what these comments are saying. Also, if we are using inline comments I believe PEP8 wants them to be two spaces after the end of the line.

File: website/notifications/views.py [hunk 36]
```diff
@@ -0,0 +1,72 @@
...
+from framework.auth.decorators import must_be_logged_in
+from model import Subscription
+from flask import request
+from modularodm import Q
+from modularodm.exceptions import NoResultsFound
+from modularodm.storage.mongostorage import KeyExistsException
+from website.notifications.constants import NOTIFICATION_TYPES
+from website.notifications import utils
+from website.project.decorators import must_be_valid_project
+
```
Comment (HarryRybacki) [inline website/notifications/views.py]: NP: two lines between imports and code

File: scripts/send_digest.py [hunk 37]
```diff
@@ -0,0 +1,113 @@
...
+            mails.send_mail(
+                to_addr=user.username,
+                mimetype='html',
+                mail=mails.DIGEST,
+                name=user.fullname,
+                message=sorted_messages,
+                url=web_url_for('user_notifications', _absolute=True),
+                callback=remove_sent_digest_notifications.s(digest_notification_ids=digest_notification_ids)
+            )
+
```
Comment (HarryRybacki) [inline scripts/send_digest.py]: NP: Two lines between top level functions

File: tests/test_notifications.py [hunk 38]
```diff
@@ -0,0 +1,841 @@
+from modularodm import Q
```
Comment (HarryRybacki) [inline tests/test_notifications.py]: NP: Imports should be grouped into system, third-party, and local source tree each of which are separated by a single empty line.

File: tests/test_notifications.py [hunk 39]
```diff
@@ -0,0 +1,841 @@
...
+            object_id=node._id,
+            event_name='comments'
+        )
+        node_subscription.save()
+
+        # Assert that user does not receive an email when someone comments on the component
+        emails.check_parent(node._id, 'comments', [])
+        assert_false(mock_send.called)
+
+
```
Comment (HarryRybacki) [inline tests/test_notifications.py]: NP: Kill this empty line

File: tests/test_notifications.py [hunk 40]
```diff
@@ -0,0 +1,841 @@
...
+    # @mock.patch('website.notifications.emails.email_transactional')
```
Comment (HarryRybacki) [inline tests/test_notifications.py]: Is this needed or can we remove it?

File: tests/test_utils.py [hunk 41]
```diff
@@ -1,15 +1,26 @@
 # -*- coding: utf-8 -*-
+import datetime
```
Comment (HarryRybacki) [inline tests/test_utils.py]: NP: Imports should be grouped into system, third-party, and local source tree each of which are separated by a single empty line.

File: website/notifications/emails.py [hunk 42]
```diff
@@ -0,0 +1,170 @@
+import pytz
```
Comment (HarryRybacki) [inline website/notifications/emails.py]: NP: Imports should be grouped into system, third-party, and local source tree each of which are separated by a single empty line.

File: website/notifications/views.py [hunk 43]
```diff
@@ -0,0 +1,72 @@
+import httplib as http
```
Comment (HarryRybacki) [inline website/notifications/views.py]: NP: Imports should be grouped into system, third-party, and local source tree each of which are separated by a single empty line.

File: website/notifications/views.py [hunk 44]
```diff
@@ -0,0 +1,72 @@
...
+from modularodm.exceptions import NoResultsFound
+from modularodm.storage.mongostorage import KeyExistsException
+from website.notifications.constants import NOTIFICATION_TYPES
+from website.notifications import utils
+from website.project.decorators import must_be_valid_project
+
+@must_be_logged_in
+def get_subscriptions(auth):
+    return utils.format_user_and_project_subscriptions(auth.user)
+
```
Comment (HarryRybacki) [inline website/notifications/views.py]: NP: Two empty lines between top level functions.

File: website/notifications/views.py [hunk 45]
```diff
@@ -0,0 +1,72 @@
...
+@must_be_valid_project
+@must_be_logged_in
+def get_node_subscriptions(auth, **kwargs):
+    node = kwargs['node'] or kwargs['project']
+    return utils.format_data(auth.user, [node._id], [])
+
```
Comment (HarryRybacki) [inline website/notifications/views.py]: NP: Two empty lines between top level functions.

File: website/profile/views.py [hunk 46]
```diff
@@ -28,7 +27,7 @@
 from website.util.sanitize import escape_html
 from website.util.sanitize import strip_html
 from website.profile import utils as profile_utils
-
+from website import settings
 
 logger = logging.getLogger(__name__)
```
Comment (HarryRybacki) [inline website/profile/views.py]: NP: Two lines between the last import and code.

File: website/profile/views.py [hunk 47]
```diff
@@ -74,6 +73,16 @@ def date_or_none(date):
         logger.exception(error)
         return None
```
Comment (HarryRybacki) [inline website/profile/views.py]: NP: Two empty lines between top level functions.

File: website/static/js/pages/project-settings-page.js [hunk 48]
```diff
@@ -84,6 +99,6 @@ $(document).ready(function() {
         }
     });
 
-
 });
 
+
```
Comment (HarryRybacki) [inline website/static/js/pages/project-settings-page.js]: NP: Remove this extra line

File: website/templates/profile/notifications.mako (hunk 30 above)
Comment (caneruguz) [inline website/templates/profile/notifications.mako]: I think this PR takes care of it: https://github.com/samanehsan/osf.io/pull/5
Saman picked the look and we tested functionality to work in save success or fail scenarios. When she merges you can take a look.

File: scripts/send_digest.py (hunk 31 above)
Comment (sloria) [inline scripts/send_digest.py]: Could you just instantiate another Celery app?

Comment (samanehsan) [inline scripts/send_digest.py]: Oh, okay!

File: website/notifications/emails.py (hunk 33 above)
Comment (sloria) [inline website/notifications/emails.py]: +1

File: tests/test_notifications.py [hunk 49]
```diff
@@ -0,0 +1,953 @@
...
+import collections
+import pytz
+from mako.lookup import Template
+from tests.base import OsfTestCase, capture_signals
+from nose.tools import *  # PEP8 asserts
+from framework.auth.signals import contributor_removed, node_deleted
+from framework.auth import Auth
+from website.util import web_url_for
+from website.notifications.model import Subscription, DigestNotification
+from website.notifications.constants import SUBSCRIPTIONS_AVAILABLE, NOTIFICATION_TYPES, USER_SUBSCRIPTIONS_AVAILABLE
```
Comment (chrisseto) [inline tests/test_notifications.py]: It would be better to use `from website.notifications import constants` and then just use `constants.USER_SU...`

File: website/templates/project/settings.mako [hunk 50]
```diff
@@ -186,13 +213,16 @@
 % endfor
 
 <%def name="javascript_bottom()">
+    <% import json %>
```
Comment (chrisseto) [inline website/templates/project/settings.mako]: Is this being used anywhere?

File: website/templates/profile/notifications.mako [hunk 51]
```diff
@@ -44,13 +45,25 @@
...
+        </div>
     </div>
 </div>
 </%def>
 
 <%def name="javascript()">
     <% import website %>
+    <% import json %>
```
Comment (chrisseto) [inline website/templates/profile/notifications.mako]: Is this used at all?

File: tests/test_views.py (hunk 34 above)
Comment (brianjgeiger) [inline tests/test_views.py]: Yeah, I think the test will go in the test_notifications.py as a test of the utility function.

**Comment by samanehsan** [comment]
@HarryRybacki, I'm saving information on what projects to show on the Node model when a user creates, updates, or removes a subscription so you will have to test with new project subscriptions or change the existing subscriptions on your current projects.

**Comment by sloria** [comment]
Merged into https://github.com/CenterForOpenScience/osf.io/pull/1955 .
